docker_cli/_shell_cmds.py

```diff
@@ -0,0 +1,369 @@
+        stringify : bool, optional
+            Whether to return a string or list - if true, will return a string.
+            Defaults to False.
+
+        Returns
+        -------
+        Union[List[str], str]
+            The commands.
+        """
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def get_name() -> str:
+        """
+        Returns the name of the command.
+
+        Returns
+        -------
+        str
+            The command name.
+        """
+        ...
+
+
+class Wget(URLReader):
+    @staticmethod
+    def generate_read_command(
+        target: str,
+        *,
+        output_target: Optional[Union[os.PathLike[str], str]] = None,
+        stringify: bool = False,
+    ) -> Union[List[str], str]:
+        retval = ["wget", target]
+        if output_target is not None:
+            retval += ["-O", str(output_target)]
+        if stringify:
+            return " ".join(retval)
+        return retval  # pragma: no cover
+
+    @staticmethod
+    def get_name() -> str:  # pragma: no cover
+        return "wget"
+
+
+class cURL(URLReader):
+    @staticmethod
+    def generate_read_command(
+        target: str,
+        *,
+        output_target: Optional[Union[os.PathLike[str], str]] = None,
+        stringify: bool = False,
+    ) -> Union[List[str], str]:
+        retval = ["curl", "--ssl", target]
+        if output_target is not None:
+            retval += ["-o", str(output_target)]
+        if stringify:
+            return " ".join(retval)
+        return retval  # pragma: no cover
+
+    @staticmethod
+    def get_name() -> str:  # pragma: no cover
+        return "curl"
+
+
+def get_package_manager(name: str) -> PackageManager:
+    """
+    Returns the PackageManager associated with a name.
+
+    Parameters
+    ----------
+    name : str
+        The name of the command
+
+    Returns
+    -------
+    PackageManager
+        An associated PackageManager object
+
+    Raises
+    ------
+    ValueError
+        When `name` does not correspond to a supported package manager.
+    """
+    name_cleaned = name.lower()
+    if name_cleaned == "apt-get":
+        return AptGet()
+    if name_cleaned == "yum":
+        return Yum()
+    else:
+        raise ValueError(f"Package manager '{name}' not recognized!")
+
+
+def get_url_reader(name: str) -> Type[URLReader]:
+    """
+    Returns the URLReader associated with a command name.
+
+    Parameters
+    ----------
+    name : str
+        The command name.
+
+    Returns
+    -------
+    URLReader
+        The associated URLReader class
+
+    Raises
+    ------
+    ValueError
+        When `name` does not correspond to a supported URL reader.
+    """
+    url_readers: Dict[str, Type[URLReader]] = {"curl": cURL, "wget": Wget}
+    if name in url_readers.keys():
+        url_reader: Type[URLReader] = url_readers[name]
+        return url_reader
+    else:
+        raise ValueError(f"URL Reader {name} not recognized!")
```

Comment:
In general, this pattern of checking whether a key exists in a dict before getting its value is inefficient because it's performing two redundant hash table lookups -- one when you check `if key in my_dict` and then again when you get `my_dict[key]`.

Instead, it's more efficient to use a `try ... except KeyError` pattern. That way, the hash function is only evaluated once.

```suggestion
    try:
        return url_readers[name]
    except KeyError:
        raise ValueError(f"URL Reader {name} not recognized!")
```